### PR TITLE
Binary

### PR DESCRIPTION
**Pull request by rbturnbull**
- Adds in more metrics
- refactors calculation of the loss
- modifications from @djmannion 's comments in the previous PR.

### CODE REVIEW

File: polytorch/plots.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            )
+            colors.extend(my_colors)
+        else:
+            labels.append(input_type.name)
+            colors.append(getattr(input_type, "color", '') or next(cmap))
+
+    weights = torch.cat(weights, dim=0).detach()
+
+    # Perform a principal component analysis
+    pca = PCA(n_components=n_components)
```
Comment (djmannion) [inline polytorch/plots.py:90]: I haven't had a chance to go through this PR properly yet, but wanted to flag something here. What I was talking about last week about 'explained variance' was related to this - the `pca` object will have an `explained_variance_ratio_` property that describes the proportion of variance that is explained by each component. I was thinking that if you ran the PCA with the number of components equal to the dimensionality of the embedding, you would be able to see that say 90% of the variance was explained by $n$ components - which might be a heuristic for a ballpark number of embedding dimensions?

File: polytorch/util.py [hunk 2]
```diff
@@ -17,7 +15,7 @@ def total_size(data_types:List[PolyData]) -> int:
     return sum(data_type.size() for data_type in data_types)
 
 
-def split_tensor(tensor:Tensor, data_types:List[PolyData], feature_axis:int=-1) -> Tuple[Tensor]:
+def split_tensor(tensor:Tensor, data_types, feature_axis:int=-1) -> Tuple[Tensor]:
```
Comment (djmannion) [inline polytorch/util.py]: Should the return type be `Tuple[Tensor, ...]` here, because it will not necessarily be just one tensor?

File: tests/test_loss.py [hunk 3]
```diff
@@ -33,6 +33,45 @@ def test_loss_categorical():
...
+def test_loss_binary():
+    target = torch.tensor([False, True, False, True, False]).unsqueeze(1)
+    prediction = (target.float() - 0.5) * 10.0
+
+    loss_fn = PolyLoss([BinaryData()])
+    loss = loss_fn(prediction, target)
+    assert loss.item() < 0.01
+
+    # change targets
+    loss = loss_fn(prediction, ~target)
```
Comment (djmannion) [inline tests/test_loss.py]: I haven't come across `~` in Python code before. Interesting that pytorch seems to be using it in what is a pretty non-intuitive way (to me) - i.e., not as bit flip.

File: polytorch/metrics.py [hunk 4]
```diff
@@ -9,38 +9,148 @@ def get_predictions_target_for_index(predictions, *targets, data_index=None, fea
...
+
+    See:
+        - https://www.sciencedirect.com/science/article/pii/S2590005619300049#bib73
+        - https://arxiv.org/pdf/1707.03237.pdf
+        - https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=1717643
+    """
+    n_classes = predictions.shape[feature_axis]
+    numerator = 0.0
+    denominator = 0.0
+    slice_indices = [slice(0, None)] * len(predictions.shape)
```
Comment (djmannion) [inline polytorch/metrics.py:81]: Interesting use of `slice` here; seems like a way to use `:` when you don't know which dimensions you want to use it on.

**Review by djmannion — APPROVED** [review]
Not a very thorough review sorry, because a lot of the content is over my head (learning pytorch is high up on my todo list). But it all seems to be structured and tested very nicely and the content makes sense to me at a surface level.

I have found it useful in my code to make `mypy --strict` pass though, might be worth considering for future. Some of it is busy work but I've also found it to pick up on some issues or even just code smell that is worth addressing.

File: polytorch/metrics.py (hunk 4 above)
Comment (rbturnbull) [reply polytorch/metrics.py:81]: that's right, that's why I used it. I don't know a better way

File: tests/test_loss.py (hunk 3 above)
Comment (rbturnbull) [reply tests/test_loss.py]: Good point. I think I'll change this to `torch.logical_not(target)` so that it is more explicit what is going on.

File: polytorch/plots.py (hunk 1 above)
Comment (rbturnbull) [reply polytorch/plots.py:90]: thanks for this suggestion. I get what you mean now! I'll give that a go.

File: polytorch/util.py (hunk 2 above)
Comment (rbturnbull) [reply polytorch/util.py]: thanks for this. I didn't properly understand python type hinting for tuples!

File: polytorch/metrics.py (hunk 4 above)
Comment (djmannion) [reply polytorch/metrics.py:81]: I think an alternative in numpy would be something like `np.take(a=predictions, indices=i, axis=feature_axis)`. But I think your approach is just as good if not better and I don't know if there is an equivalent of `np.take` in pytorch (maybe [`index_select`](https://pytorch.org/docs/stable/generated/torch.index_select.html#torch.index_select)?).

**Comment by rbturnbull** [comment]
Thanks for the comments @djmannion - I'll add mypy as a dev dependency on another branch and work to resolve all the issues it comes up with.
